File: SwrveSDK/SDK/Messaging/SwrveButtonTheme.m
#import "SwrveButtonTheme.h"

@implementation SwrveButtonTheme

@synthesize fontSize;
@synthesize fontFile;
@synthesize fontDigest;
@synthesize fontNativeStyle;
@synthesize fontPostscriptName;
@synthesize topPadding;
@synthesize rightPadding;
@synthesize bottomPadding;
@synthesize leftPadding;
@synthesize cornerRadius;
@synthesize fontColor;
@synthesize bgColor;
@synthesize borderWidth;
@synthesize borderColor;
@synthesize bgImage;
@synthesize truncate;
@synthesize pressedState;
@synthesize focusedState;
@synthesize hAlign;

- (id)initWithDictionary:(NSDictionary *)themeData {
    if (self = [super init]) {
        self.fontSize = [themeData objectForKey:@"font_size"];
        self.fontFile = [themeData objectForKey:@"font_file"];
        if ([themeData objectForKey:@"font_digest"]) {
            self.fontDigest = [themeData objectForKey:@"font_digest"];
        }
        if ([themeData objectForKey:@"font_native_style"]) {
            self.fontNativeStyle = [themeData objectForKey:@"font_native_style"];
        }
        self.fontPostscriptName = [themeData objectForKey:@"font_postscript_name"];
        NSDictionary *padding = [themeData objectForKey:@"padding"];
        self.topPadding = [padding objectForKey:@"top"];
        self.bottomPadding = [padding objectForKey:@"bottom"];
        self.rightPadding = [padding objectForKey:@"right"];
        self.leftPadding = [padding objectForKey:@"left"];
        self.cornerRadius = [themeData objectForKey:@"corner_radius"];
        self.fontColor = [themeData objectForKey:@"font_color"];
        self.fontDigest = [themeData objectForKey:@"font_digest"];
        if ([themeData objectForKey:@"bg_color"]) {
            self.bgColor = [themeData objectForKey:@"bg_color"];
        }
        if ([themeData objectForKey:@"border_width"]) {
            self.borderWidth = [themeData objectForKey:@"border_width"];
        }
        if ([themeData objectForKey:@"border_color"]) {
            self.borderColor = [themeData objectForKey:@"border_color"];
        }
        if ([themeData objectForKey:@"bg_image"] && [themeData objectForKey:@"bg_image"] != [NSNull null]) {
            self.bgImage = [(NSDictionary *) [themeData objectForKey:@"bg_image"] objectForKey:@"value"];
        }
        self.truncate = [[themeData objectForKey:@"truncate"] boolValue];
        self.pressedState = [[SwrveButtonThemeState alloc] initWithDictionary:[themeData objectForKey:@"pressed_state"]];
        if ([themeData objectForKey:@"focused_state"] && [themeData objectForKey:@"focused_state"] != [NSNull null]) {
            self.focusedState = [[SwrveButtonThemeState alloc] initWithDictionary:[themeData objectForKey:@"focused_state"]];
        }
        self.hAlign = [themeData objectForKey:@"h_align"];
    }
    return self;
}

@end

File: SwrveSDKSwift/SwrveButtonTheme.swift
import Foundation

@objc public class SwrveButtonTheme: NSObject {

    // MARK: - Properties

    @objc public var fontSize: NSNumber?
    @objc public var fontFile: String?
    @objc public var fontDigest: String?
    @objc public var fontNativeStyle: String?
    @objc public var fontPostscriptName: String?

    @objc public var topPadding: NSNumber?
    @objc public var rightPadding: NSNumber?
    @objc public var bottomPadding: NSNumber?
    @objc public var leftPadding: NSNumber?

    @objc public var cornerRadius: NSNumber?

    @objc public var fontColor: String?
    @objc public var bgColor: String?

    @objc public var borderWidth: NSNumber?
    @objc public var borderColor: String?

    @objc public var bgImage: String?
    @objc public var truncate: Bool = false

    @objc public var pressedState: SwrveButtonThemeState
    @objc public var focusedState: SwrveButtonThemeState?

    @objc public var hAlign: String?

    // MARK: - Init

    @objc public init(dictionary themeData: [String: Any]) {
        // Simple values
        fontSize = SwrveButtonTheme.cgFloat(from: themeData["font_size"]) as NSNumber?
        fontFile = themeData["font_file"] as? String

        if let digest = themeData["font_digest"] as? String {
            fontDigest = digest
        }
        if let nativeStyle = themeData["font_native_style"] as? String {
            fontNativeStyle = nativeStyle
        }

        fontPostscriptName = themeData["font_postscript_name"] as? String

        // Padding
        if let padding = themeData["padding"] as? [String: Any] {
            topPadding = SwrveButtonTheme.cgFloat(from: padding["top"]) as NSNumber?
            bottomPadding = SwrveButtonTheme.cgFloat(from: padding["bottom"]) as NSNumber?
            rightPadding = SwrveButtonTheme.cgFloat(from: padding["right"]) as NSNumber?
            leftPadding = SwrveButtonTheme.cgFloat(from: padding["left"]) as NSNumber?
        }

        cornerRadius = SwrveButtonTheme.cgFloat(from: themeData["corner_radius"]) as NSNumber?
        fontColor = themeData["font_color"] as? String

        // Note: original Obj-C sets fontDigest again; harmless no-op here if present
        if let digestAgain = themeData["font_digest"] as? String {
            fontDigest = digestAgain
        }

        if let bg = themeData["bg_color"] as? String {
            bgColor = bg
        }
        if let bw = themeData["border_width"] {
            borderWidth = SwrveButtonTheme.cgFloat(from: bw) as NSNumber?
        }
        if let bc = themeData["border_color"] as? String {
            borderColor = bc
        }

        // bg_image = { "value": "<string>" } (and may be NSNull)
        if let bgImageAny = themeData["bg_image"], !(bgImageAny is NSNull),
            let bgImageDict = bgImageAny as? [String: Any]
        {
            bgImage = bgImageDict["value"] as? String
        }

        truncate = (themeData["truncate"] as? Bool) ?? false

        // States
        let pressedDict = (themeData["pressed_state"] as? [String: Any]) ?? [:]
        self.pressedState = SwrveButtonThemeState(dictionary: pressedDict)

        if let focusedAny = themeData["focused_state"], !(focusedAny is NSNull),
            let focusedDict = focusedAny as? [String: Any]
        {
            self.focusedState = SwrveButtonThemeState(dictionary: focusedDict)
        }

        hAlign = themeData["h_align"] as? String
    }

    // MARK: - Helpers

    private static func cgFloat(from any: Any?) -> CGFloat? {
        switch any {
        case let n as NSNumber:
            return CGFloat(truncating: n)
        case let d as Double:
            return CGFloat(d)
        case let f as Float:
            return CGFloat(f)
        case let i as Int:
            return CGFloat(i)
        case let s as String:
            if let d = Double(s) { return CGFloat(d) }
            return nil
        default:
            return nil
        }
    }
}
